# Linear Regression with PyTorch on the MPG Dataset

Attempt will be made to build a simple linear regression model using the auto MPG dataset. 


I will select some features of the dataset to work with namely, number of cylinders, displacement, horsepower, weight and acceleration as the numeric (i.e. continuous) features with the manufacturing origin as an unordered categorical feature with three possible discrete values, 1,2 and 3 which correspond to US, Europe and Japan respectively. Model year is an ordered categorical (ordinal) feature.

In [26]:
import pandas as pd
import sklearn
import sklearn.model_selection
import torch
import torch.nn as nn
from torch.nn.functional import one_hot
from torch.utils.data import DataLoader,Dataset,TensorDataset


The initial step involves loading the data and executing essential preprocessing tasks. These tasks encompass the removal of incomplete rows, the division of the dataset into training and testing subsets, and the standardization of continuous features.

In [27]:
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data'
column_names = ['MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight',
                'Acceleration', 'Model Year', 'Origin']

df = pd.read_csv(url, names=column_names,
                 na_values = "?", comment='\t',
                 sep=" ", skipinitialspace=True)

df.tail()


MPG  Cylinders  Displacement  Horsepower  Weight  Acceleration  \
393  27.0          4         140.0        86.0  2790.0          15.6   
394  44.0          4          97.0        52.0  2130.0          24.6   
395  32.0          4         135.0        84.0  2295.0          11.6   
396  28.0          4         120.0        79.0  2625.0          18.6   
397  31.0          4         119.0        82.0  2720.0          19.4   

     Model Year  Origin  
393          82       1  
394          82       2  
395          82       1  
396          82       1  
397          82       1

In [28]:
# dropping nas 
print(df.isna().sum())

df = df.dropna()
df = df.reset_index(drop=True)
df.tail()


MPG             0
Cylinders       0
Displacement    0
Horsepower      6
Weight          0
Acceleration    0
Model Year      0
Origin          0
dtype: int64


MPG  Cylinders  Displacement  Horsepower  Weight  Acceleration  \
387  27.0          4         140.0        86.0  2790.0          15.6   
388  44.0          4          97.0        52.0  2130.0          24.6   
389  32.0          4         135.0        84.0  2295.0          11.6   
390  28.0          4         120.0        79.0  2625.0          18.6   
391  31.0          4         119.0        82.0  2720.0          19.4   

     Model Year  Origin  
387          82       1  
388          82       2  
389          82       1  
390          82       1  
391          82       1

In [29]:
# splitting the data into train and test sets
df_train, df_test = sklearn.model_selection.train_test_split(df, train_size=0.8, random_state=1)
# Making a dataframe containing the stats
train_stats = df_train.describe().transpose()
train_stats


count         mean         std     min     25%     50%     75%  \
MPG           313.0    23.404153    7.666909     9.0    17.5    23.0    29.0   
Cylinders     313.0     5.402556    1.701506     3.0     4.0     4.0     8.0   
Displacement  313.0   189.512780  102.675646    68.0   104.0   140.0   260.0   
Horsepower    313.0   102.929712   37.919046    46.0    75.0    92.0   120.0   
Weight        313.0  2961.198083  848.602146  1613.0  2219.0  2755.0  3574.0   
Acceleration  313.0    15.704473    2.725399     8.5    14.0    15.5    17.3   
Model Year    313.0    75.929712    3.675305    70.0    73.0    76.0    79.0   
Origin        313.0     1.591054    0.807923     1.0     1.0     1.0     2.0   

                 max  
MPG             46.6  
Cylinders        8.0  
Displacement   455.0  
Horsepower     230.0  
Weight        5140.0  
Acceleration    24.8  
Model Year      82.0  
Origin           3.0

In [30]:
# standardizing continuous features
numeric_column_names = ['Cylinders', 'Displacement', 'Horsepower', 'Weight', 'Acceleration']

df_train_norm, df_test_norm = df_train.copy(), df_test.copy()

for col_name in numeric_column_names:
    mean = train_stats.loc[col_name, 'mean']
    std  = train_stats.loc[col_name, 'std']
    df_train_norm.loc[:, col_name] = (df_train_norm.loc[:, col_name] - mean)/std
    df_test_norm.loc[:, col_name] = (df_test_norm.loc[:, col_name] - mean)/std
    
df_train_norm.tail()


MPG  Cylinders  Displacement  Horsepower    Weight  Acceleration  \
203  28.0  -0.824303     -0.901020   -0.736562 -0.950031      0.255202   
255  19.4   0.351127      0.413800   -0.340982  0.293190      0.548737   
72   13.0   1.526556      1.144256    0.713897  1.339617     -0.625403   
235  30.5  -0.824303     -0.891280   -1.053025 -1.072585      0.475353   
37   14.0   1.526556      1.563051    1.636916  1.470420     -1.359240   

     Model Year  Origin  
203          76       3  
255          78       1  
72           72       1  
235          77       1  
37           71       1

Next, we bucket the model year categories as follows:

In [31]:
boundaries = torch.tensor([73, 76, 79])
 
v = torch.tensor(df_train_norm['Model Year'].values)
df_train_norm['Model Year Bucketed'] = torch.bucketize(v, boundaries, right=True)

v = torch.tensor(df_test_norm['Model Year'].values)
df_test_norm['Model Year Bucketed'] = torch.bucketize(v, boundaries, right=True)

numeric_column_names.append('Model Year Bucketed')


In [32]:
# Checking the numeric column names
numeric_column_names

['Cylinders',
 'Displacement',
 'Horsepower',
 'Weight',
 'Acceleration',
 'Model Year Bucketed']

We then use one-hot encoding on the origin feature to convert it to into the dense format and also complete the train and test tensors

In [33]:
total_origin = len(set(df_train_norm['Origin']))

origin_encoded = one_hot(torch.from_numpy(df_train_norm['Origin'].values) % total_origin)
x_train_numeric = torch.tensor(df_train_norm[numeric_column_names].values)
x_train = torch.cat([x_train_numeric, origin_encoded], 1).float()
 
origin_encoded = one_hot(torch.from_numpy(df_test_norm['Origin'].values) % total_origin)
x_test_numeric = torch.tensor(df_test_norm[numeric_column_names].values)
x_test = torch.cat([x_test_numeric, origin_encoded], 1).float()


Then, create the label tensors from the ground truth MPG values:

In [34]:
y_train = torch.tensor(df_train_norm['MPG'].values).float()
y_test = torch.tensor(df_test_norm['MPG'].values).float()


## Training a DNN Regression Model

We will first create a data loader that uses a batch size of 8 for the train data:

In [35]:
train_ds = TensorDataset(x_train, y_train)
batch_size = 8
torch.manual_seed(1)
train_dl = DataLoader(train_ds, batch_size, shuffle=True)



Then, we build a model with two fully connected layers, one having eight units and the other having 4 units. 

In [36]:
hidden_units = [8, 4]
input_size = x_train.shape[1]

all_layers = []
for hidden_unit in hidden_units:
    layer = nn.Linear(input_size, hidden_unit)
    all_layers.append(layer)
    all_layers.append(nn.ReLU())
    input_size = hidden_unit

all_layers.append(nn.Linear(hidden_units[-1], 1))

model = nn.Sequential(*all_layers)

model


Sequential(
  (0): Linear(in_features=9, out_features=8, bias=True)
  (1): ReLU()
  (2): Linear(in_features=8, out_features=4, bias=True)
  (3): ReLU()
  (4): Linear(in_features=4, out_features=1, bias=True)
)

Having defined the model, we also define the loss function, which should be MSE for regression and use stochastic gradient descent for optimization.

In [37]:
loss_fn = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.001)


We then train the model for 200 epochs, displaying the train loss every 20 epochs

In [38]:
torch.manual_seed(1)
num_epochs = 200
log_epochs = 20 
for epoch in range(num_epochs):
    loss_hist_train = 0
    for x_batch, y_batch in train_dl:
        pred = model(x_batch)[:, 0]
        loss = loss_fn(pred, y_batch)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        loss_hist_train += loss.item()
    if epoch % log_epochs==0:
        print(f'Epoch {epoch}  Loss {loss_hist_train/len(train_dl):.4f}')


Epoch 0  Loss 536.1047


Epoch 20  Loss 8.4361
Epoch 40  Loss 7.8695
Epoch 60  Loss 7.1891
Epoch 80  Loss 6.7062
Epoch 100  Loss 6.7599
Epoch 120  Loss 6.3124
Epoch 140  Loss 6.6864
Epoch 160  Loss 6.7648
Epoch 180  Loss 6.2156


After 200 epochs, the train loss was around 6 from the initial of about 540. This means that the model fit the data well. 

We can then move on to evaluate the regression performance of the trained model on the test dataset. To predict the target values on new data points, we can feed their features to the model.

In [39]:
with torch.no_grad():
    pred = model(x_test.float())[:, 0]
    loss = loss_fn(pred, y_test)
    print(f'Test MSE: {loss.item():.4f}')
    print(f'Test MAE: {nn.L1Loss()(pred, y_test).item():.4f}')


Test MSE: 9.6130
Test MAE: 2.1211


The MSE on the test set is 9.6 and the MAE is 2.1. This not too far off from that of the train set, showing that the model can generalize with a little work. 